modules/r2dbc/src/main/java/org/testcontainers/r2dbc/ConnectionPublisher.java

```diff
@@ -0,0 +1,152 @@
+     * - "synchronized" is used to avoid races
+     * - Reactive Streams spec violations are not checked (e.g. non-positive request)
+     */
+    private class StateMachineSubscription implements Subscription {
+
+        private final Subscriber<? super Connection> actual;
+
+        Subscription subscriptionState;
+
+        StateMachineSubscription(Subscriber<? super Connection> actual) {
+            this.actual = actual;
+            subscriptionState = new WaitRequestSubscriptionState();
+        }
+
+        @Override
+        public synchronized void request(long n) {
+            subscriptionState.request(n);
+        }
+
+        @Override
+        public synchronized void cancel() {
+            subscriptionState.cancel();
+        }
+
+        synchronized void transitionTo(SubscriptionState newState) {
+            subscriptionState = newState;
+            newState.enter();
+        }
+
+        abstract class SubscriptionState implements Subscription {
+            void enter() {
+            }
+        }
+
+        class WaitRequestSubscriptionState extends SubscriptionState {
+
+            @Override
+            public void request(long n) {
+                transitionTo(new WaitFutureCompletionSubscriptionState());
+            }
+
+            @Override
+            public void cancel() {
+            }
+        }
+
+        class WaitFutureCompletionSubscriptionState extends SubscriptionState {
+
+            private CompletableFuture<ConnectionFactory> future;
+
+            @Override
+            void enter() {
+                this.future = futureSupplier.get();
+
+                future.whenComplete((connectionFactory, e) -> {
+                    if (e != null) {
+                        actual.onSubscribe(EmptySubscription.INSTANCE);
+                        actual.onError(e);
+                        return;
+                    }
+
+                    Publisher<? extends Connection> publisher = connectionFactory.create();
+                    transitionTo(new ProxySubscriptionState(publisher));
+                });
+            }
+
+            @Override
+            public void request(long n) {
+            }
+
+            @Override
+            public void cancel() {
+                future.cancel(true);
+            }
+        }
+
+        class ProxySubscriptionState extends SubscriptionState implements Subscriber<Connection> {
+
+            private final Publisher<? extends Connection> publisher;
+
+            private Subscription s;
+
+            ProxySubscriptionState(Publisher<? extends Connection> publisher) {
+                this.publisher = publisher;
+            }
+
+            @Override
+            void enter() {
+                publisher.subscribe(this);
+            }
+
+            @Override
+            public void request(long n) {
+                // Ignore
+            }
+
+            @Override
+            public void cancel() {
+                s.cancel();
```

Comment:
might be a case when 

```
future just completed ->
void enter() { publisher.subscribe(this); } executed but subscription may be delivered later (See https://github.com/reactive-streams/reactive-streams-jvm/issues/486 for the clarification)
so the next cancel which is waiting on the synchronized block will end-up with NPE.
```

The simples solution for that is to add flag and synchronization between request cancel and onsubscriber to ensure we are not ending up with NPE if the subscription has not arrived yet

Reply:
good catch